sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala

```diff
@@ -29,40 +29,31 @@ import org.apache.spark.sql.catalyst.rules.Rule
 
 /**
  * Make sure that a view's child plan produces the view's output attributes. We wrap the child
- * with a Project and add an alias for each output attribute. The attributes are resolved by
- * name. This should be only done after the batch of Resolution, because the view attributes are
- * not completely resolved during the batch of Resolution.
+ * with a Project and add an alias for each output attribute by mapping the child output by index,
+ * if the view output doesn't have the same number of columns with the child output, throw an
+ * AnalysisException.
+ * This should be only done after the batch of Resolution, because the view attributes are not
+ * completely resolved during the batch of Resolution.
  */
 case class AliasViewChild(conf: CatalystConf) extends Rule[LogicalPlan] {
   override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
     case v @ View(_, output, child) if child.resolved =>
-      val resolver = conf.resolver
-      val newOutput = output.map { attr =>
-        val originAttr = findAttributeByName(attr.name, child.output, resolver)
-        // The dataType of the output attributes may be not the same with that of the view output,
-        // so we should cast the attribute to the dataType of the view output attribute. If the
-        // cast can't perform, will throw an AnalysisException.
-        Alias(Cast(originAttr, attr.dataType), attr.name)(exprId = attr.exprId,
-          qualifier = attr.qualifier, explicitMetadata = Some(attr.metadata))
+      if (output.length != child.output.length) {
+        throw new AnalysisException(
+          s"The view output ${output.mkString("[", ",", "]")} doesn't have the same number of " +
+            s"columns with the child output ${child.output.mkString("[", ",", "]")}")
+      }
+      val newOutput = output.zip(child.output).map {
+        case (attr, originAttr) =>
+          if (attr.dataType != originAttr.dataType) {
```

Comment:
Seems that Hive supports UpCast between child output and view output, for example:
```
hive> create table testtable as select 1 a, 2L b;
hive> create view testview as select * from testtable;
hive> select * from testview;
OK
1	2
Time taken: 0.11 seconds, Fetched: 1 row(s)
hive> alter table testtable change column a a bigint;
hive> alter table testtable change column b b string;
hive> desc testtable;
OK
a                   	bigint              	                    
b                   	string              	                    
Time taken: 0.15 seconds, Fetched: 2 row(s)
hive> desc testview;
OK
a                   	int                 	                    
b                   	bigint              	                    
Time taken: 0.038 seconds, Fetched: 2 row(s)
hive> select * from testview;
OK
1	2
Time taken: 0.172 seconds, Fetched: 1 row(s)
```